src/components/Clipboard.react.js

```diff
@@ -0,0 +1,147 @@
+import React, {Component} from 'react'; // eslint-disable-line no-unused-vars
+import PropTypes from 'prop-types';
+import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
+import {faCopy, faCheckCircle} from '@fortawesome/free-regular-svg-icons';
+
+const clipboardAPI = navigator.clipboard;
+
+/**
+ * The Clipboard component copies text to the clipboard
+ */
+
+export default class Clipboard extends React.Component {
+    constructor(props) {
+        super(props);
+        this.copyToClipboard = this.copyToClipboard.bind(this);
+        this.copySuccess = this.copySuccess.bind(this);
+        this.getText = this.getText.bind(this);
+        this.stringifyId = this.stringifyId.bind(this);
+        this.state = {
+            copied: false,
+        };
+    }
+
+    // stringifies object ids used in pattern matching callbacks
+    stringifyId(id) {
+        if (typeof id !== 'object') {
+            return id;
+        }
+        const stringifyVal = v => (v && v.wild) || JSON.stringify(v);
+        const parts = Object.keys(id)
+            .sort()
+            .map(k => JSON.stringify(k) + ':' + stringifyVal(id[k]));
+        return '{' + parts.join(',') + '}';
+    }
+
+    copySuccess() {
+        this.props.setProps({
+            n_clicks: this.props.n_clicks + 1,
+        });
+
+        const showCopiedIcon = 1000;
+        this.setState({copied: true}, () => {
+            setTimeout(() => {
+                this.setState({copied: false});
+            }, showCopiedIcon);
+        });
+    }
+
+    getText() {
+        // get the inner text.  If none, use the content of the value param
+        const id = this.stringifyId(this.props.target_id);
+        const target = document.getElementById(id);
```

Comment:
what if you provide a bad `target_id`? 
```
document.getElementById('asdf').innerText
> Uncaught TypeError: Cannot read property 'innerText' of null
```
perhaps add something like `if (!target) { return ''; }` immediately after the `getElementById`?

Reply:
Right - that handles the error, but it could be really hard to debug.  Is there a way to display a message in the error box?

Reply:
Ah, that's a good question - no, I don't know of a good way to expose this error to the dash devtools error box - it's happening in an event handler, and React _does_ wrap these so it's possible there's a way we could access them from the renderer. Right now though we don't have a way to do that.

OK, so let's leave it as you have it now - that'll at least prevent `copySuccess` and display the error in the js console, and later we can investigate whether there's some generic way to trap event handler errors.

I suppose if you think `Uncaught TypeError: Cannot read property 'innerText' of null` is too removed from the issue of not having the right ID, we could also explicitly raise such an error:
```
if (!target) {
    throw new Error("Clipboard copy failed: no element found for target_id " + target_id);
}
```

Reply:
Just belatedly passing through and noting this issue for exposing errors in Dash DevTools: https://github.com/plotly/dash/issues/1088